torch_geometric/loader/link_neighbor_loader.py

```diff
@@ -4,20 +4,56 @@
 from torch import Tensor
 
 from torch_geometric.data import Data, HeteroData
+from torch_geometric.data.feature_store import FeatureStore
+from torch_geometric.data.graph_store import EdgeLayout, GraphStore
 from torch_geometric.loader.base import DataLoaderIterator
 from torch_geometric.loader.neighbor_loader import NeighborSampler
-from torch_geometric.loader.utils import filter_data, filter_hetero_data
+from torch_geometric.loader.utils import (
+    filter_data,
+    filter_feature_store,
+    filter_hetero_data,
+)
 from torch_geometric.typing import InputEdges, NumNeighbors, OptTensor
 
 
 class LinkNeighborSampler(NeighborSampler):
-    def __init__(self, data, *args, neg_sampling_ratio: float = 0.0, **kwargs):
+    def __init__(
+        self,
+        data,
+        *args,
+        neg_sampling_ratio: float = 0.0,
+        num_src_nodes: Optional[int] = None,
+        num_dst_nodes: Optional[int] = None,
+        **kwargs,
+    ):
         super().__init__(data, *args, **kwargs)
         self.neg_sampling_ratio = neg_sampling_ratio
 
-        if issubclass(self.data_cls, Data):
+        if self.data_cls == 'custom':
+            _, graph_store = data
+            edge_attrs = graph_store.get_all_edge_attrs()
+            edge_types = [attr.edge_type for attr in edge_attrs]
+
+            # Edge label index is part of the graph:
+            if self.input_type in edge_types:
+                self.num_src_nodes, self.num_dst_nodes = edge_attrs[
+                    edge_types.index(self.input_type)].size
+
+            # Otherwise:
+            elif num_src_nodes is None or num_dst_nodes is None:
```

Comment:
`FeatureStore.size` is kind of broken since we currently need to provide an attribute to get the size, which is not available in general, which is why this approach was chosen :( Happy to refactor once we clean up size.